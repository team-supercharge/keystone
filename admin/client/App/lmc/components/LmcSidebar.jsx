import React, { Component } from 'react'
import PropTypes from 'prop-types'
import _ from 'lodash'
import LmcSidebarItem from './LmcSidebarItem.jsx'
import LmcSidebarFilter from './LmcSidebarFilter.jsx'
import LmcCreateButton from './LmcCreateButton.jsx'

export default class LmcSidebar extends Component {
    state = {
        nameFilter: '',
        displayActiveItems: true
    }

    componentDidMount () {
        const { items, selectedItem, setSelectedItem } = this.props;
        let shownItems = _.filter(items, (item) => !this.calculateHidden(item))
        if (shownItems.length && !selectedItem) {
            setSelectedItem(shownItems[0].id)
        }
    }

    calculateHidden = (item) => {
        const { first, last } = item.name
        return (
            !`${first} ${last}`.match(new RegExp(this.state.nameFilter, 'i')) || 
            (this.state.displayActiveItems && item.status !== 'active')
        )
    }

    handleFormChange = (e) => {
        this.setState({
            [e.target.name]: e.target.value
        })
    }

    handleSwitchChange = () => {
        this.setState(prevState => ({ 
            displayActiveItems: !prevState.displayActiveItems 
        }))
    }

    renderTitle() {
        const { title } = this.props
        if (!title) return null

        return (
            <div style={styles.titleContainer}>
                <span style={styles.title}>
                    { title }
                </span>
            </div>
        )
    }

    render() {
        const { 
            itemLabel,
            listId,
            items, 
            selectedItem, 
            setSelectedItem,
            onCreate,
        } = this.props;
        
        let shownItems = _.filter(items, (item) => !this.calculateHidden(item))

        return (
            <div className='lmc-sidebar'>
                { this.renderTitle() }
                <LmcSidebarFilter
                    onFormChange={this.handleFormChange}
                    onSwitchChange={this.handleSwitchChange}
                    isChecked={!this.state.displayActiveItems}
                />
                <LmcCreateButton
                    buttonText={itemLabel}
                    listId={listId}
                    title={`Add a new ${itemLabel}`}
                    onCreate={onCreate}
                    style={styles.button}
                />
                <ul className='lmc-sidebar-list'>
                        { shownItems.map((item, index) => {
                            return (
                                <LmcSidebarItem 
                                    key={index}
                                    itemData={item}
                                    onClick={() => setSelectedItem(item.id)}
                                    isSelected={item.id === selectedItem}
                                />
                            )
                        }) }
                </ul>
            </div>
        )
    }
}

const styles = {
    button: {
        borderRadius: 0,
    },
    titleContainer: {
        backgroundColor: 'white',
        padding: '50px 0px 20px 0px',
        fontSize: 20,
        fontWeight: 600,
        textAlign: 'center',
    }
};

LmcSidebar.propTypes = {
    itemLabel: PropTypes.string.isRequired,
    listId: PropTypes.string.isRequired,
    items: PropTypes.array,
    onCreate: PropTypes.func.isRequired,
    selectedItem: PropTypes.string,
    setSelectedItem: PropTypes.func.isRequired,
    title: PropTypes.string,
};