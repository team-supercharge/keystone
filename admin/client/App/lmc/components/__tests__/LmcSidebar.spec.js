jest.mock('../../../shared/CreateForm')

import React from 'react'
import { shallow } from 'enzyme'
import LmcSidebar from '../LmcSidebar.jsx'

describe('LmcSidebar', () => {
    let wrapper
    let items
    const onCreateMock = jest.fn()
    const setSelectedItemMock = jest.fn()

    beforeEach(() => {
        items = [
            { id: 'testId1', name: { first: 'test', last: 'id1' }, status: 'active' },
            { id: 'testId2', name: { first: 'test', last: 'id2' }, status: 'active' },
            { id: 'testId3', name: { first: 'test', last: 'id3' }, status: 'inactive' },
        ]
        wrapper = shallow(
            <LmcSidebar
                itemLabel='Item'
                listId='Item'
                items={items}
                onCreate={onCreateMock}
                setSelectedItem={setSelectedItemMock}
                selectedItem={items[0].id}
                title='TestTitle'
            />
        )
    })

    it('should render correctly', () => {
        expect(wrapper).toMatchSnapshot()
    })

    it('renders an unordered list', () => {
        expect(wrapper.find('ul').length).toEqual(1)
    })

    it('renders a number of sidebar items based on its active items list', () => {
        expect(wrapper.find('LmcSidebarItem').length).toEqual(2)
    })

    it('renders the sidebar filter', () => {
        expect(wrapper.find('LmcSidebarFilter').length).toEqual(1)
    })

    it('renders a LmcCreateButton with the correct props', () => {
        const button = wrapper.find('LmcCreateButton')

        expect(button.props().listId).toEqual('Item')
        expect(button.props().title).toEqual('Add a new Item')
        expect(button.props().onCreate).toEqual(onCreateMock)
    })
})