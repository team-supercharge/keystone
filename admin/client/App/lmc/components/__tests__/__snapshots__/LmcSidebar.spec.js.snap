// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LmcSidebar should render correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="lmc-sidebar"
  >
    <LmcSidebarFilter
      isChecked={false}
      onFormChange={[Function]}
      onSwitchChange={[Function]}
    />
    <LmcCreateButton
      buttonText="Item"
      listId="Item"
      onCreate={[MockFunction]}
      style={
        Object {
          "borderRadius": 0,
        }
      }
      title="Add a new Item"
    />
    <ul
      className="lmc-sidebar-list"
    >
      <LmcSidebarItem
        isSelected={true}
        itemData={
          Object {
            "id": "testId1",
            "name": Object {
              "first": "test",
              "last": "id1",
            },
            "status": "active",
          }
        }
        onClick={[Function]}
      />
      <LmcSidebarItem
        isSelected={false}
        itemData={
          Object {
            "id": "testId2",
            "name": Object {
              "first": "test",
              "last": "id2",
            },
            "status": "active",
          }
        }
        onClick={[Function]}
      />
    </ul>
  </div>,
  "nodes": Array [
    <div
      className="lmc-sidebar"
    >
      <LmcSidebarFilter
        isChecked={false}
        onFormChange={[Function]}
        onSwitchChange={[Function]}
      />
      <LmcCreateButton
        buttonText="Item"
        listId="Item"
        onCreate={[MockFunction]}
        style={
          Object {
            "borderRadius": 0,
          }
        }
        title="Add a new Item"
      />
      <ul
        className="lmc-sidebar-list"
      >
        <LmcSidebarItem
          isSelected={true}
          itemData={
            Object {
              "id": "testId1",
              "name": Object {
                "first": "test",
                "last": "id1",
              },
              "status": "active",
            }
          }
          onClick={[Function]}
        />
        <LmcSidebarItem
          isSelected={false}
          itemData={
            Object {
              "id": "testId2",
              "name": Object {
                "first": "test",
                "last": "id2",
              },
              "status": "active",
            }
          }
          onClick={[Function]}
        />
      </ul>
    </div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <LmcSidebar
        itemLabel="Item"
        items={
          Array [
            Object {
              "id": "testId1",
              "name": Object {
                "first": "test",
                "last": "id1",
              },
              "status": "active",
            },
            Object {
              "id": "testId2",
              "name": Object {
                "first": "test",
                "last": "id2",
              },
              "status": "active",
            },
            Object {
              "id": "testId3",
              "name": Object {
                "first": "test",
                "last": "id3",
              },
              "status": "inactive",
            },
          ]
        }
        listId="Item"
        onCreate={[MockFunction]}
        selectedItem="testId1"
        setSelectedItem={[MockFunction]}
      />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": LmcSidebar {
        "_reactInternalInstance": [Circular],
        "calculateHidden": [Function],
        "context": Object {},
        "handleFormChange": [Function],
        "handleSwitchChange": [Function],
        "props": Object {
          "itemLabel": "Item",
          "items": Array [
            Object {
              "id": "testId1",
              "name": Object {
                "first": "test",
                "last": "id1",
              },
              "status": "active",
            },
            Object {
              "id": "testId2",
              "name": Object {
                "first": "test",
                "last": "id2",
              },
              "status": "active",
            },
            Object {
              "id": "testId3",
              "name": Object {
                "first": "test",
                "last": "id3",
              },
              "status": "inactive",
            },
          ],
          "listId": "Item",
          "onCreate": [MockFunction],
          "selectedItem": "testId1",
          "setSelectedItem": [MockFunction],
        },
        "refs": Object {},
        "state": Object {
          "displayActiveItems": true,
          "nameFilter": "",
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="lmc-sidebar"
        >
          <LmcSidebarFilter
            isChecked={false}
            onFormChange={[Function]}
            onSwitchChange={[Function]}
          />
          <LmcCreateButton
            buttonText="Item"
            listId="Item"
            onCreate={[MockFunction]}
            style={
              Object {
                "borderRadius": 0,
              }
            }
            title="Add a new Item"
          />
          <ul
            className="lmc-sidebar-list"
          >
            <LmcSidebarItem
              isSelected={true}
              itemData={
                Object {
                  "id": "testId1",
                  "name": Object {
                    "first": "test",
                    "last": "id1",
                  },
                  "status": "active",
                }
              }
              onClick={[Function]}
            />
            <LmcSidebarItem
              isSelected={false}
              itemData={
                Object {
                  "id": "testId2",
                  "name": Object {
                    "first": "test",
                    "last": "id2",
                  },
                  "status": "active",
                }
              }
              onClick={[Function]}
            />
          </ul>
        </div>,
        "_debugID": 2,
        "_renderedOutput": <div
          className="lmc-sidebar"
        >
          <LmcSidebarFilter
            isChecked={false}
            onFormChange={[Function]}
            onSwitchChange={[Function]}
          />
          <LmcCreateButton
            buttonText="Item"
            listId="Item"
            onCreate={[MockFunction]}
            style={
              Object {
                "borderRadius": 0,
              }
            }
            title="Add a new Item"
          />
          <ul
            className="lmc-sidebar-list"
          >
            <LmcSidebarItem
              isSelected={true}
              itemData={
                Object {
                  "id": "testId1",
                  "name": Object {
                    "first": "test",
                    "last": "id1",
                  },
                  "status": "active",
                }
              }
              onClick={[Function]}
            />
            <LmcSidebarItem
              isSelected={false}
              itemData={
                Object {
                  "id": "testId2",
                  "name": Object {
                    "first": "test",
                    "last": "id2",
                  },
                  "status": "active",
                }
              }
              onClick={[Function]}
            />
          </ul>
        </div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <LmcSidebar
    itemLabel="Item"
    items={
      Array [
        Object {
          "id": "testId1",
          "name": Object {
            "first": "test",
            "last": "id1",
          },
          "status": "active",
        },
        Object {
          "id": "testId2",
          "name": Object {
            "first": "test",
            "last": "id2",
          },
          "status": "active",
        },
        Object {
          "id": "testId3",
          "name": Object {
            "first": "test",
            "last": "id3",
          },
          "status": "inactive",
        },
      ]
    }
    listId="Item"
    onCreate={[MockFunction]}
    selectedItem="testId1"
    setSelectedItem={[MockFunction]}
  />,
}
`;
